chain/messagepool/pruning.go

```diff
@@ -0,0 +1,311 @@
+
+		from := msg.Message.From
+
+		getBalStart := build.Clock.Now()
+		if _, ok := inclNonces[from]; !ok {
+			act, err := mp.api.StateGetActor(from, nil)
+			if err != nil {
+				log.Warnf("failed to check message sender balance, skipping message: %+v", err)
+				continue
+			}
+
+			inclNonces[from] = act.Nonce
+			inclBalances[from] = act.Balance
+		}
+		getbal += build.Clock.Since(getBalStart)
+
+		if inclBalances[from].LessThan(msg.Message.RequiredFunds()) {
+			tooLowFundMsgs++
+			// todo: drop from mpool
+			continue
+		}
+
+		if msg.Message.Nonce > inclNonces[from] {
+			tooHighNonceMsgs++
+			continue
+		}
+
+		if msg.Message.Nonce < inclNonces[from] {
+			continue
+		}
+
+		inclNonces[from] = msg.Message.Nonce + 1
+		inclBalances[from] = types.BigSub(inclBalances[from], msg.Message.RequiredFunds())
+		sm := outBySender[from]
+		if sm == nil {
+			sm = &senderMeta{
+				lastReward: big.Zero(),
+			}
+		}
+
+		sm.gasLimit = append(sm.gasLimit, sm.lastGasLimit+msg.Message.GasLimit)
+		sm.lastGasLimit = sm.gasLimit[len(sm.gasLimit)-1]
+
+		guessGasStart := build.Clock.Now()
+		guessedGas, err := gasguess.GuessGasUsed(ctx, types.EmptyTSK, msg, al)
+		guessGasDur += build.Clock.Since(guessGasStart)
+		if err != nil {
+			log.Infow("failed to guess gas", "to", msg.Message.To, "method", msg.Message.Method, "err", err)
+		}
+
+		estimatedReward := big.Mul(types.NewInt(uint64(guessedGas)), msg.Message.GasPrice)
+
+		sm.gasReward = append(sm.gasReward, big.Add(sm.lastReward, estimatedReward))
+		sm.lastReward = sm.gasReward[len(sm.gasReward)-1]
+
+		sm.msgs = append(sm.msgs, msg)
+
+		outBySender[from] = sm
+	}
+
+	gasLimitLeft := int64(build.BlockGasLimit)
+
+	orderedSenders := make([]address.Address, 0, len(outBySender))
+	for k := range outBySender {
+		orderedSenders = append(orderedSenders, k)
+	}
+	sort.Slice(orderedSenders, func(i, j int) bool {
+		return bytes.Compare(orderedSenders[i].Bytes(), orderedSenders[j].Bytes()) == -1
+	})
+
+	out := make([]*types.SignedMessage, 0, build.BlockMessageLimit)
+	{
+		for {
+			var bestSender address.Address
+			var nBest int
+			var bestGasToReward float64
+
+			// TODO: This is O(n^2)-ish, could use something like container/heap to cache this math
+			for _, sender := range orderedSenders {
+				meta, ok := outBySender[sender]
+				if !ok {
+					continue
+				}
+				for n := range meta.msgs {
+					if meta.gasLimit[n] > gasLimitLeft {
+						break
+					}
+
+					if n+len(out) > build.BlockMessageLimit {
+						break
+					}
+
+					gasToReward, _ := new(big2.Float).SetInt(meta.gasReward[n].Int).Float64()
+					gasToReward /= float64(meta.gasLimit[n])
+
+					if gasToReward >= bestGasToReward {
+						bestSender = sender
+						nBest = n + 1
+						bestGasToReward = gasToReward
+					}
+				}
+			}
+
+			if nBest == 0 {
+				break // block gas limit reached
+			}
+
+			{
+				out = append(out, outBySender[bestSender].msgs[:nBest]...)
+				gasLimitLeft -= outBySender[bestSender].gasLimit[nBest-1]
```

Comment:
via this you limit messages to one block gas limit
we will end up reaching `gasLimitLeft ~= 0` and iterating through all senders trying to add more messages but we will always fail in for loop above

Reply:
yeap, fixing now